text/0000-target-tier-policy.md

```diff
@@ -0,0 +1,222 @@
+  concerned with that target. Rust developers may be expected to not
+  gratuitously break a tier 2 target, but are not expected to become experts in
+  every tier 2 target, and are not expected to provide target-specific
+  implementations for every tier 2 target.
+- Where possible, tier 2 targets should provide documentation for the Rust
+  community for how to build and run tests for the platform, ideally using
+  emulation.
+- The target development team should not only fix target-specific issues, but
+  should use any such issue as an opportunity to educate the Rust community
+  about portability to their target, and enhance their documentation of the
+  target.
+- The target must build reliably in CI.
+- Building the target must not take substantially longer than other targets.
+- Tier 2 targets must support building on the existing targets used for CI
+  infrastructure.
+- Tier 2 targets must not impose burden on the authors of pull requests, or
+  other developers in the community, to ensure that tests pass for the target.
+  In particular, do not post comments (automated or manual) on a PR that
+  suggests a block on the PR based on tests failing for the target. (A PR
+  author must not break the build of a tier 2 target, but need not ensure the
+  tests pass for a tier 2 target, even if notified that they fail.)
+- The target development team should regularly run the testsuite for the
+  target, and should fix any test failures in a reasonably timely fashion.
+- A tier 2 target may be demoted or removed if it no longer meets these
+  requirements. Any proposal for demotion or removal will be CCed to people who
+  have previously worked on the platform, and will be communicated widely to
+  the Rust community before being dropped from a stable release.
+- All tier 3 requirements apply.
+
+Note: some tier 2 platforms additionally have binaries built to run on them as
+a host (such as `rustc` and `cargo`). Such a platform must meet all the
+requirements above, and must additionally get the compiler and infrastructure
+team to approve the building of host tools.
+
+## Tier 1 target policy
+
+At this tier, the Rust project guarantees that a target builds and passes all
+tests, and will reject patches that fail to build or pass the testsuite on a
+target. We hold tier 1 targets to our highest standard of requirements.
+
+- Any new tier 1 target requires compiler team approval based on these
+  requirements.
+- In addition, the infrastructure team must approve the integration of the
+  target into CI, and the CI-related requirements. This review and approval
+  will typically take place in the PR adding the target to CI.
+- In addition, the release team must approve the long-term viability of the
+  target, and the additional work of supporting the target.
+- Tier 1 targets must have substantial, widespread interest within the
+  developer community, and must serve the ongoing needs of multiple production
```

Comment:
What are the risks associated with a high profile (i.e., well known and well funded) user of Rust needing support for a target that (at least in the short/medium term) is not deemed useful by the wider community? Does this risk a split of the ecosystem or an exit of that user from the Rust community? Presumably targets can be de-facto tier 1 without having official status, and it would be up to the high profile user to work with the community to ensure the target meets their needs without inconveniencing wider Rust development.

Reply:
Do you have anything specific in mind here?

Reply:
> What are the risks associated with a high profile (i.e., well known and well funded) user of Rust needing support for a target that (at least in the short/medium term) is not deemed useful by the wider community

Would you expand on what you would mean by support? It's not likely that the Rust community would implement and maintain platform that they don't deem useful, a user is however free maintain a platform themselves and submit patches to Rust (as along as they also follow the rules for that tier).

> Does this risk a split of the ecosystem or an exit of that user from the Rust community?

I don't think platform support alone is controversial enough for a split or exit. It would require other external factors as well. (e.g. providing a level of support within a tight timeline, or politics.) Is your concern that to developing the target might cause too much churn for Rust developers to review? If this was the case it wouldn't be unreasonable for the user to have a fork for them to work on, and upstream patches at a longer pace (kind of like how rust treats LLVM) until the platform is more stable in rustc, though I suspect this wouldn't be needed in reality.

Reply:
Let's take Fuchsia as an example. This is unlikely to be important to anyone outside of Google in the near time. Are there risks associated with Google not having wider community support for that target?

Reply:
@rylev That's a bit of an open-ended question. =P

Reply:
@rylev Well I think firstly it's important to know whether you the user want the platform to be a host or target platform, e.g. Do you want developers to be able to build and run their code on the platform itself or is it satisfactory if they only compile to that platform and use another for development? A target platform doesn't need support everything a host platform does, so it would be lower impact. At the cost of not being able to develop on the platform itself, which may be fine depending on your use case.

With this there's the risk of having to put in the initial work investment and continued maintenance to ensure that the platform continues to compile, build, and work for the end software being built with new compilers. This cost increases with the amount of tooling you want to offer on the host platform and how much you want the platform to be tier 1 quality. There's also the work to implement the standard library for your platform if it's radically different than others.

Reply:
I agree that this requirement is vague. Who decides "substantial" / "widespread" / "interest", etc?

Reply:
I've elaborated substantially on this.

I don't think the requirement should (or can) become fully objective. But I've called this out as a subjective requirement, and discussed how it's evaluated.